lib/as/notifications/fanout.rb

```diff
@@ -15,8 +15,8 @@ def initialize
         super
       end
 
-      def subscribe(pattern = nil, block = Proc.new)
-        subscriber = Subscribers.new pattern, block
+      def subscribe(pattern = nil, callback = nil, &block)
+        subscriber = Subscribers.new pattern, callback || block
```

Comment:
I think you can get away without `callback` at all – `block` will be `nil` when it needs to be?

Reply:
@rgarner I tried that first but it didn't work:

```
Error: test_publish_with_subscriber(Notifications::SyncPubSubTest): ArgumentError: wrong number of arguments (given 2, expected 0..1)
/code/lib/as/notifications/fanout.rb:18:in `subscribe'
/code/test/notifications_test.rb:164:in `test_publish_with_subscriber'
     161: 
     162:     def test_publish_with_subscriber
     163:       subscriber = TestSubscriber.new
  => 164:       @notifier.subscribe nil, subscriber
     165:       @notifier.publish :foo
     166: 
     167:       assert_equal [[:foo]], subscriber.publishes
/usr/local/bundle/bundler/gems/rails-e17e25cd23e8/activesupport/lib/active_support/testing/setup_and_teardown.rb:72:in `block in run'
/usr/local/bundle/bundler/gems/rails-e17e25cd23e8/activesupport/lib/active_support/callbacks.rb:403:in `_run__2729736072672214470__setup__4306434224111393863__callbacks'
/usr/local/bundle/bundler/gems/rails-e17e25cd23e8/activesupport/lib/active_support/callbacks.rb:405:in `__run_callback'
/usr/local/bundle/bundler/gems/rails-e17e25cd23e8/activesupport/lib/active_support/callbacks.rb:385:in `_run_setup_callbacks'
/usr/local/bundle/bundler/gems/rails-e17e25cd23e8/activesupport/lib/active_support/callbacks.rb:81:in `run_callbacks'
/usr/local/bundle/bundler/gems/rails-e17e25cd23e8/activesupport/lib/active_support/testing/setup_and_teardown.rb:70:in `run'
```

Reply:
Ok, will clone and check it out. I'm still uncertain as to the changed behaviours in Ruby 3.0.0 :)

Reply:
@rgarner Thanks for testing! :pray:

I checked with the latest upstream code, and they are using something similar. So I guess it's okay. :smile: 

https://github.com/rails/rails/blob/2f012f74ccdb42ea88363484770dd6cd5e82fc93/activesupport/lib/active_support/notifications/fanout.rb#L24-L25

Reply:
Ah, ok, my confusion arises from the fact that [`Subscribers.new` will take](https://github.com/bernd/as-notifications/blob/a1be83ddfbad17d8acbcbbcb47d4c70b51bb6f34/lib/as/notifications/fanout.rb#L62-L66) either an object responding to `:start` and `:finish`, or a `Proc` (and transform those to `Evented` or `Timed` subscribers respectively, where `Timed` just does a `#call`). In its own naming convention, that's a `listener`. 

So what you have is right, but maybe we could call it `listener` instead of `callback`? Then it's consistent with the `Subscribers` terminology and avoids the implication that `callback` might just be a `Proc` (when it can be an object responding to `:start` and `:finish`).

Reply:
Ah, my massive comment crossed over with yours. So they have `callable` ... well ok :)

Reply:
@rgarner No worries. :smile: I will make it `callable` to make a future code sync easier.

Reply:
I really do miss types sometimes. Nearly all of the hints about type information leak into the names of things...